File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:        
        def valid(mid):
            curr=0
            for p in piles: curr += math.ceil(p/mid)
            return curr<=h
        
        l, r, = 1, max(piles)
        ans = r
        while l<=r:
            mid = (l+r)//2
            if valid(mid):
                ans = min(ans, mid)
                r=mid-1
            else:
                l=mid+1
        return ans

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int l=0, res=0, r=0;
        for(int p: piles) r=Math.max(r, p);
        while(l<=r){
            int mid = l+(r-l)/2;
            if(valid(mid, h, piles)){
                res=mid;
                r=mid-1;
            }
            else l=mid+1;
        }
        return res;
    }
    private boolean valid(int mid, int h, int[] piles){
        long time=0;
        for(int p: piles) time += Math.ceil((double)p / (double)mid);
        return time<=h;
    }
}